### PR TITLE
[Core][Projects] Get packaging types from build plugins

### PR DESCRIPTION
**Pull request by kjsmita6**
This is an improved version of #538. Instead of creating an extension point, the project will now look through all of its build/extension plugins, open the jar files, and find/parse components.xml (as suggested by @laeubi ). 

Two things I am unsure of:
1. I'm not super knowledgeable in Maven/m2e internals. I know enough about it, but not as much as might be required. I tested this code with my own product's custom packaging types and it was able to load all of them into the pom editor from components.xml. I'm not sure if everyone has it set up the same way. If not, this probably won't work.
2. I don't think m2e can use any plexus XML parsing code/if there is a way to parse components.xml already. If there is I can remove my own parsing code in favor of an existing one.

Again, please let me know if anything can be changed/improved. I am happy to have feedback and fix whatever is necessary. This is something we really need in our product so I would really like to have it included in m2e.

### CODE REVIEW

**Comment by laeubi** [comment]
> 1\. I'm not sure if everyone has it set up the same way. If not, this probably won't work.

I think it don't has to work for everyone, we can improve it later if there are other means of defining this. Also with #546 we might be able to leverage other techniques if full extension support is available.

> 2\. I don't think m2e can use any plexus XML parsing code/if there is a way to parse components.xml already.

I think standard java xml parsing is fine and might even be more flexible

File: org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java
Comment (laeubi) [inline org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java:357]: Maybe it would be worth here to use a TreeSet<String> to first collect all packing types (initialized with the hard-coded defaults) so there is no need for filtering and no possible duplicates.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (laeubi) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:81]: I think this is not necessary if we make the UI collecting everything in a Set first (see comment below).

**Review by laeubi — COMMENTED** [review]
I think this is fine, just one minor adjustment.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (mickaelistria) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:79]: Let's drop it and directly use `ArtifactHandler.ROLE` from Maven API

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (mickaelistria) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:77]: I'm annoyed I didn't find constant in Plexus or Maven for that...

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (mickaelistria) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:81]: Ideally, the solution shouldn't require such exception and would be generic enough to work without them. (It's not a blocker for a merge though)

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java
Comment (mickaelistria) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: Since this is only useful for the UI (OverviewPage), I don't think we should make it an API here, and instead we should probably just move the code to OverviewPage directly.

**Review by mickaelistria — CHANGES_REQUESTED** [review]
It's great progress.
However, all that should be moved to Overview page as it's not really useful as part of the IMavenProjectFacade in general and we prefer keeping such important API minimal and more focused.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java
Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: I think `getAvailablePackagingTypes()` is good here but  `updateAvailablePackagingTypes()` should not be part of the API.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:77]: I even don't found the place where it is parsed ...

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java
Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java:81]: See my suggestions, I think the overview page should simply use a Set<String> and we should not bother here....

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java
Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: Why I think its useful in the API is that I would expect in the future that this information is read via the maven-extension mechanism and then we don't want to remember the Overviewpage uses some kind of own solution.
Also we probably should check elsewhere in the code if we have hard-coded defaults that better call this method instead.

Comment (mickaelistria) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: > Why I think its useful in the API is that I would expect in the future that this information is read via the maven-extension mechanism and then we don't want to remember the Overviewpage uses some kind of own solution.

That's not how API are supposed to be growing. Some code is promoted as an API if there is a clear need from it, from multiple consumers and if it makes sense to host the logic for those clients into the object because of some strong coupling. It's not the case here. IMavenProjectFacade represents one project and its info. The available types is most often useless for a project and not really related to the project but more to the classrealm, the set is moveable so it's unclear from API what to expect as value... It could be some utility class somewhere, but definitely not a getter in IMavenProjectFacade.
At this stage, there is no good reason to make it API, so we should avoid future pain and not make it API, at least not that way.
The more API != the better.

**Comment by laeubi** [comment]
@mickaelistria I think we should include this in 1.20, there is not much to be left, it is just a bit moving things here or there ... and it is local enough to have low risk for regression.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java
Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: Then lets move it to OverviewPage but include it in 1.20 ...

**Comment by laeubi** [comment]
@kjsmita6 do you think you can move the code to the OverviewPage and update your PR as of today?

**Comment by mickaelistria** [comment]
It's not ready and I don't want to wait further with the release. I see a bad trend of the release scope growing everyday. At some point, if we want 2.0 to happen, we need to make 1.20 highest priority.
If we wait for this one to be complete, then another interesting change will come up, and we'll then wait for this other interesting change and so on. It's already been so for the 3 last weeks. We don't have 3 more weeks to waste if we want 1.20 in next SimRel.

**Comment by laeubi** [comment]
> If we wait for this one to be complete, then another interesting change will come up, and we'll then wait for this other interesting change and so on. It's already been so for the 3 last weeks. We don't have 3 more weeks to waste if we want 1.20 in next SimRel.

Didn't we say we don't want to be coupled to SimRel time-schedule? :-)
I don't say we should include other poping up but this one.

**Comment by akurtakov** [comment]
We don't want to be coupled as there are many other things to release at end of cycle thus Mickael tried to release it earlier but it ended up with endless feature growth. If 1.20 don't make it 2022-03 there is no point in even releasing it as 2022-06 will have 2.0 and many(most?) users will not get 1.20 at all.

**Comment by laeubi** [comment]
> as 2022-06 will have 2.0

Sorry of being pessimistic but I don't see any guarantee that we will have 2.0 in 2022-06 as its completely unclear how far we got so maybe 2022-06 will still ship 1.20, that's why I have completed some important features (especially for supporting pom-dependencies) and those where people otherwise would complain (conflicting mappings).

And I don't see any 'endless' growth' there is a clear decrease in PRs targeting for 1.20 and waiting for this where a user has actually invest time for a completely different solution to came up with a important improvement for his usecase sound fair and waiting for at least another 3 Months until it bubbles up in the IDE seems not very profitable.

**Comment by akurtakov** [comment]
2022-06 should ship 2.0 - it's a matter of getting rid of all old stuff that's planned and get 2.0 out. Doable even in April IMHO. Any feature addition that don't make it should be for 2.1,... It's important to deliver faster rather than prolong releases.

**Comment by laeubi** [comment]
> It's important to deliver faster rather than prolong releases.

But that's impossible with current eclipse-rules ... anyways I don't get why we repel first-time contributors due to delay their contribution because of a unclear "feeling" delaying the 1.20 at one or two days won't harm if we simply agree not including anything else.

If someone is eager to start right now, one could simply do it on a branch and we merge it in after the release, this change is local enough to be not affected by any API changes.

**Comment by laeubi** [comment]
> Mickael asked for 1.20 release January 21st - a release should have been done in the next few days.

Yeah bust asking for a release is not enough, I asked for a planed date and it was said we should not plan a date, its a bit late to complain now that there is a date (now) but that was never communicated ... Also one should check tickets and assign milestones, the code was simply not ready at this point of time to be released,

> someone else could have picked the releng had and do 1.20.1 or 1.21 in two weeks and repeated even one more time if needed.

someone sometimes aka never ... I just can strongly encourage to include this as it is clear and small and valuable and unlikely we will ever do a 1.20.1 once we started with the 2.0 stuff. Anyways of course no one is oblique to follow this advice, its just makes me sad that we often have this 'If I'm not affected I don't care' habit. For me https://github.com/eclipse-m2e/m2e-core/issues/549 was a clear blocker as we would otherwise would get instant complains from m2e users once this is released to a wider audience. The second one was https://github.com/eclipse-m2e/m2e-core/issues/543 what will be crucial if we want to support maven locations in the upcoming 2022-06 and I can't really find anything else that was done for the 1.20 in the meantime so its really a bit unfair to claim we are adding "endless new features" here...

**Comment by laeubi** [comment]
> 2022-06 should ship 2.0

Could you then please add the latest date we should have the 2.0 available[ here as a due-date](https://github.com/eclipse-m2e/m2e-core/milestones/5/edit)?

**Comment by akurtakov** [comment]
> > 2022-06 should ship 2.0
> 
> Could you then please add the latest date we should have the 2.0 available[ here as a due-date](https://github.com/eclipse-m2e/m2e-core/milestones/5/edit)?

Done. May 24th is the latest date at which release should be public in order to make it to M3 as per https://wiki.eclipse.org/Category:SimRel-2022-06 . Wed is last contribution day in these weeks. Thu is aggregation. Fri is going public.

**Comment by laeubi** [comment]
> > > 2022-06 should ship 2.0
> > 
> > 
> > Could you then please add the latest date we should have the 2.0 available[ here as a due-date](https://github.com/eclipse-m2e/m2e-core/milestones/5/edit)?
> 
> Done. May 24th is the latest date at which release should be public in order to make it to M3 as per https://wiki.eclipse.org/Category:SimRel-2022-06 . Wed is last contribution day in these weeks. Thu is aggregation. Fri is going public.

Thanks, I have for now set Friday for 1.20 please adjust if that seems unrealistic.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java
Comment (kjsmita6) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/project/IMavenProjectFacade.java:172]: Moved to overview page.

File: org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java
Comment (laeubi) [inline org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java:376]: I think you could simply place the code from updateAvailablePackagingTypes() right here and safe the field `packagingTypes` then and use a local variable instead.

**Review by laeubi — CHANGES_REQUESTED** [review]
Just a minor adjustment.

File: org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java
Comment (kjsmita6) [reply org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java:376]: Have this done but just waiting for final review from @mickaelistria so I can just push all changes at once.

Comment (laeubi) [reply org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java:376]: Just squash your commits and push them here, so we can see the current status for this

Comment (mickaelistria) [reply org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java:376]: +1 with @laeubi here, please push your progress without waiting for anyone. FWIW, I'm fine with this new patch, and would also like to see the changes suggested by @laeubi implemented. After that, I think it's find to merge as it's an isolated change bringing a great feature.

Comment (kjsmita6) [reply org.eclipse.m2e.editor/src/org/eclipse/m2e/editor/pom/OverviewPage.java]: Alright, pushed my change. Everything should be here now.

**Review by mickaelistria — APPROVED** [review]
That's a great improvement to me! Thanks a lot.
@laeubi OK to merge?

**Review by laeubi — APPROVED** [review]

**Pull request merged**
